File: apps/api/signalyze_api/engine/signals/rating_polarity.py
"""Twin of ``signals/rating-polarity.ts``."""

from __future__ import annotations

from collections.abc import Sequence

from signalyze_api.engine.data_files import THRESHOLDS
from signalyze_api.engine.mathutil import ramp, round6
from signalyze_api.engine.prepare import PreparedReview
from signalyze_api.models import SignalResult

T = THRESHOLDS.rating_polarity


def rating_polarity(reviews: Sequence[PreparedReview]) -> SignalResult:
    """Share of 1-star and 5-star ratings among all ratings."""
    n = len(reviews)
    five = 0
    one = 0
    for r in reviews:
        if r.review.rating == 5:
            five += 1
        elif r.review.rating == 1:
            one += 1
    share = (five + one) / n
    return SignalResult(
        id="rating_polarity",
        unusualness=round6(ramp(share, T.low, T.high)),
        value=round6(share),
        details={
            "share5": round6(five / n),
            "share1": round6(one / n),
            "shareMid": round6((n - five - one) / n),
        },
        available=True,
    )

File: packages/signals/src/signals/rating-polarity.ts
import type { SignalResult } from '@signalyze/shared';
import thresholds from '../../data/thresholds.json';
import { ramp, round6 } from '../math';
import type { PreparedReview } from '../prepare';

const T = thresholds.rating_polarity;

/** Share of 1-star and 5-star ratings among all ratings (U-shaped distributions score high). */
export function ratingPolarity(reviews: readonly PreparedReview[]): SignalResult {
  const n = reviews.length;
  let five = 0;
  let one = 0;
  for (const r of reviews) {
    if (r.review.rating === 5) five += 1;
    else if (r.review.rating === 1) one += 1;
  }
  const share = (five + one) / n;
  return {
    id: 'rating_polarity',
    unusualness: round6(ramp(share, T.low, T.high)),
    value: round6(share),
    details: {
      share5: round6(five / n),
      share1: round6(one / n),
      shareMid: round6((n - five - one) / n),
    },
    available: true,
  };
}
